### ISSUE TITLE
UART sensors not communicating on ev3dev-buster

### ISSUE THREAD

**Issue by JakubVanek** [ev3dev/ev3dev#1433]
**Describe the bug**
UART sensors are getting detected (port mode changes), but it seems that no actual communication takes place (`sensor<i>` does not get registered). EV3 touch sensor seems to work on the other hand.

**To Reproduce**
1. Download SD card image referenced below
2. Copy the image to the SD card using dd + enable 456 MHz overclock (but the bug occurs even without it) + copy `TIInit_6.2.31.bts` into /lib/firmware/ti-connectivity
3. Insert the SD card to the brick
4. Start the brick
5. Attach an EV3 sensor
6. See dmesg

The same thing happens after installing locally built ev3 kernel as well.

```
Please paste a simple program to reproduce the problem here!
```

**Expected behavior**
EV3 UART sensors work as expected.

**Screenshots**
dmesg output:
```
[    0.000000] Booting Linux on physical CPU 0x0
[    0.000000] Linux version 4.19.85-ev3dev-3-beta2-ev3-dataage+ (kuba@oggy-kuba) (gcc version 8.3.0 (GNU Toolchain for the A-profile Architecture 8.3-2019.03 (arm-rel-8.36))) #1 PREEMPT Sun Aug 9 23:09:40 CEST 2020
[    0.000000] CPU: ARM926EJ-S [41069265] revision 5 (ARMv5TEJ), cr=0005317f
[    0.000000] CPU: VIVT data cache, VIVT instruction cache
[    0.000000] OF: fdt: Machine model: LEGO MINDSTORMS EV3
[    0.000000] Memory policy: Data cache writeback
[    0.000000] Reserved memory: created DMA memory pool at 0x80000000, size 0 MiB
[    0.000000] OF: reserved mem: initialized node serial@80000000, compatible id shared-dma-pool
[    0.000000] DaVinci da850/omap-l138/am18x variant 0x1
[    0.000000] On node 0 totalpages: 16384
[    0.000000]   DMA zone: 128 pages used for memmap
[    0.000000]   DMA zone: 0 pages reserved
[    0.000000]   DMA zone: 16384 pages, LIFO batch:3
[    0.000000] random: get_random_bytes called from start_kernel+0x88/0x414 with crng_init=0
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0 
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 16256
[    0.000000] Kernel command line:   console=tty1 quiet
[    0.000000] Dentry cache hash table entries: 8192 (order: 3, 32768 bytes)
[    0.000000] Inode-cache hash table entries: 4096 (order: 2, 16384 bytes)
[    0.000000] Memory: 51020K/65536K available (5546K kernel code, 279K rwdata, 1816K rodata, 184K init, 155K bss, 14516K reserved, 0K cma-reserved)
[    0.000000] Virtual kernel memory layout:
                   vector  : 0xffff0000 - 0xffff1000   (   4 kB)
                   fixmap  : 0xffc00000 - 0xfff00000   (3072 kB)
                   vmalloc : 0xc4800000 - 0xff800000   ( 944 MB)
                   lowmem  : 0xc0000000 - 0xc4000000   (  64 MB)
                   modules : 0xbf000000 - 0xc0000000   (  16 MB)
                     .text : 0x(ptrval) - 0x(ptrval)   (5548 kB)
                     .init : 0x(ptrval) - 0x(ptrval)   ( 184 kB)
                     .data : 0x(ptrval) - 0x(ptrval)   ( 280 kB)
                      .bss : 0x(ptrval) - 0x(ptrval)   ( 156 kB)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] rcu: Preemptible hierarchical RCU implementation.
[    0.000000] 	Tasks RCU enabled.
[    0.000000] NR_IRQS: 325
[    0.000000] clocksource: timer0_1: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 79635851949 ns
[    0.000000] sched_clock: 32 bits at 24MHz, resolution 41ns, wraps every 89478484971ns
[    0.000674] Console: colour dummy device 80x30
[    0.000726] console [tty1] enabled
[    0.000843] Calibrating delay loop... 226.09 BogoMIPS (lpj=1130496)
[    0.070121] pid_max: default: 32768 minimum: 301
[    0.070720] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.070769] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.072984] CPU: Testing write buffer coherency: ok
[    0.076450] Setting up static identity map for 0xc0008400 - 0xc000843c
[    0.077048] rcu: Hierarchical SRCU implementation.
[    0.080438] devtmpfs: initialized
[    0.104523] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.104597] futex hash table entries: 256 (order: -1, 3072 bytes)
[    0.105126] pinctrl core: initialized pinctrl subsystem
[    0.108390] NET: Registered protocol family 16
[    0.109746] DMA: preallocated 256 KiB pool for atomic coherent allocations
[    0.113264] cpuidle: using governor menu
[    0.143711] gpiochip_add_data_with_key: GPIOs 0..143 (davinci_gpio.0) failed to register, -517
[    0.191032] edma 1e30000.edma: memcpy is disabled
[    0.202267] edma 1e30000.edma: TI EDMA DMA engine driver
[    0.206689] SCSI subsystem initialized
[    0.207366] usbcore: registered new interface driver usbfs
[    0.207714] usbcore: registered new interface driver hub
[    0.208094] usbcore: registered new device driver usb
[    0.211304] Advanced Linux Sound Architecture Driver Initialized.
[    0.213060] Bluetooth: Core ver 2.22
[    0.213302] NET: Registered protocol family 31
[    0.213334] Bluetooth: HCI device and connection manager initialized
[    0.213394] Bluetooth: HCI socket layer initialized
[    0.213435] Bluetooth: L2CAP socket layer initialized
[    0.213609] Bluetooth: SCO socket layer initialized
[    0.215458] clocksource: Switched to clocksource timer0_1
[    0.265372] NET: Registered protocol family 2
[    0.268301] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 4096 bytes)
[    0.268399] TCP established hash table entries: 1024 (order: 0, 4096 bytes)
[    0.268466] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)
[    0.268523] TCP: Hash tables configured (established 1024 bind 1024)
[    0.268886] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    0.268959] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    0.269701] NET: Registered protocol family 1
[    0.270997] Unpacking initramfs...
[    0.715622] random: fast init done
[    1.944006] Freeing initrd memory: 5792K
[    1.948134] Initialise system trusted keyrings
[    1.948902] workingset: timestamp_bits=30 max_order=14 bucket_order=0
[    2.006147] Key type asymmetric registered
[    2.006184] Asymmetric key parser 'x509' registered
[    2.006268] io scheduler noop registered (default)
[    2.006297] io scheduler mq-deadline registered
[    2.006319] io scheduler kyber registered
[    2.008801] da8xx-usb-phy 1c1417c.chip-controller:usb-phy: Failed to get usb1_clk48
[    2.012656] pinctrl-single 1c14120.pinmux: 160 pins, size 80
[    2.019386] ehrpwm 1f00000.pwm: Failed to get tbclk
[    2.020123] ehrpwm 1f02000.pwm: Failed to get tbclk
[    2.024818] Serial: 8250/16550 driver, 3 ports, IRQ sharing disabled
[    2.032216] 1d0c000.serial: ttyS1 at MMIO 0x1d0c000 (irq = 53, base_baud = 8250000) is a TI DA8xx/66AK2x
[    2.036341] 1d0d000.serial: ttyS2 at MMIO 0x1d0d000 (irq = 61, base_baud = 8250000) is a TI DA8xx/66AK2x
[    2.037146] serial serial0: tty port ttyS2 registered
[    2.055627] loop: module loaded
[    2.057727] zram: Added device: zram0
[    2.063627] ohci_hcd: USB 1.1 'Open' Host Controller (OHCI) Driver
[    2.063742] ohci-da8xx: DA8XX
[    2.064899] usbcore: registered new interface driver usb-storage
[    2.067221] i2c /dev entries driver
[    2.069289] davinci-wdt 1c21000.wdt: heartbeat 60 sec
[    2.074702] hidraw: raw HID events driver (C) Jiri Kosina
[    2.075321] usbcore: registered new interface driver usbhid
[    2.075345] usbhid: USB HID core driver
[    2.076015] usbcore: registered new interface driver r8712u
[    2.104119] NET: Registered protocol family 10
[    2.107957] Segment Routing with IPv6
[    2.108299] sit: IPv6, IPv4 and MPLS over IPv4 tunneling driver
[    2.110719] NET: Registered protocol family 17
[    2.110830] Bluetooth: BNEP (Ethernet Emulation) ver 1.3
[    2.110882] Bluetooth: BNEP socket layer initialized
[    2.114051] Loading compiled-in X.509 certificates
[    2.140336] GPIO line 6 (batt_volt_en) hogged as output/high
[    2.140436] GPIO line 5 (bt_clock_en) hogged as input
[    2.140548] GPIO line 51 (bt_pic_en) hogged as output/low
[    2.140649] GPIO line 78 (bt_pic_rst) hogged as output/high
[    2.140739] GPIO line 87 (bt_pic_cts) hogged as input
[    2.144851] edma 1c00000.edma: memcpy is disabled
[    2.156250] edma 1c00000.edma: TI EDMA DMA engine driver
[    2.162404] at24 0-0050: 16384 byte 24c128 EEPROM, read-only, 0 bytes/write
[    2.169094] 1c42000.serial: ttyS0 at MMIO 0x1c42000 (irq = 25, base_baud = 14250000) is a TI DA8xx/66AK2x
[    2.175011] adc ref: supplied by vcc5v
[    2.175911] ads7950 spi0.3: Linked as a consumer to regulator.1
[    2.179789] spi_davinci 1c41000.spi: Controller at 0x(ptrval)
[    2.187367] [drm] Initialized st7586 1.0.0 20170801 for spi1.0 on minor 0
[    2.619359] Console: switching to colour frame buffer device 44x21
[    2.621518] st7586 spi1.0: fb0: DRM emulated frame buffer device
[    2.621773] spi_davinci 1f0e000.spi: Controller at 0x(ptrval)
[    2.622795] ohci-da8xx 1e25000.usb: Linked as a consumer to regulator.2
[    2.623027] ohci-da8xx 1e25000.usb: USB Host Controller
[    2.624087] ohci-da8xx 1e25000.usb: new USB bus registered, assigned bus number 1
[    2.624661] ohci-da8xx 1e25000.usb: irq 59, io mem 0x01e25000
[    2.690524] usb usb1: New USB device found, idVendor=1d6b, idProduct=0001, bcdDevice= 4.19
[    2.690575] usb usb1: New USB device strings: Mfr=3, Product=2, SerialNumber=1
[    2.690603] usb usb1: Product: USB Host Controller
[    2.690634] usb usb1: Manufacturer: Linux 4.19.85-ev3dev-3-beta2-ev3-dataage+ ohci_hcd
[    2.690660] usb usb1: SerialNumber: 1e25000.usb
[    2.692925] hub 1-0:1.0: USB hub found
[    2.693193] hub 1-0:1.0: 1 port detected
[    2.696568] usb_phy_generic usb_phy_generic.0.auto: usb_phy_generic.0.auto supply vcc not found, using dummy regulator
[    2.696878] usb_phy_generic usb_phy_generic.0.auto: Linked as a consumer to regulator.0
[    2.730290] davinci_mmc 1c40000.mmc: Got CD GPIO
[    2.757287] davinci_mmc 1c40000.mmc: Using DMA, 4-bit mode
[    2.762354] i2c-legoev3 i2c-legoev3.3: registered on input port 1
[    2.764135] i2c-legoev3 i2c-legoev3.4: registered on input port 2
[    2.766187] i2c-legoev3 i2c-legoev3.5: registered on input port 3
[    2.768426] i2c-legoev3 i2c-legoev3.6: registered on input port 4
[    2.772400] lego-port port0: Registered 'ev3-ports:in1' on 'ev3-ports:in1'.
[    2.776666] lego-port port1: Registered 'ev3-ports:in2' on 'ev3-ports:in2'.
[    2.780214] lego-port port2: Registered 'ev3-ports:in3' on 'ev3-ports:in3'.
[    2.784096] lego-port port3: Registered 'ev3-ports:in4' on 'ev3-ports:in4'.
[    2.789148] lego-port port4: Registered 'ev3-ports:outA' on 'ev3-ports:outA'.
[    2.793549] lego-port port5: Registered 'ev3-ports:outB' on 'ev3-ports:outB'.
[    2.793994] mmc0: host does not support reading read-only switch, assuming write-enable
[    2.796473] mmc0: new high speed SDHC card at address 59b4
[    2.799886] mmcblk0: mmc0:59b4 SD    7.36 GiB 
[    2.804276] lego-port port6: Registered 'ev3-ports:outC' on 'ev3-ports:outC'.
[    2.807659]  mmcblk0: p1 p2
[    2.813480] lego-port port7: Registered 'ev3-ports:outD' on 'ev3-ports:outD'.
[    2.815382] ehrpwm 1f00000.pwm: PM domain ti,da850-psc1: ehrpwm will not be powered off
[    2.818542] input: LEGO MINDSTORMS EV3 Speaker as /devices/platform/sound/input/input0
[    2.828443] remoteproc remoteproc0: pru1 is available
[    2.830533] ti_omapl_pru_suart 1c30000.serial: assigned reserved memory node serial@80000000
[    2.830969] 1c30000.serial: ttySU0 at MMIO 0x1d00000 (irq = 252, base_baud = 8250000) is a suart_tty
[    2.833329] 1c30000.serial: ttySU1 at MMIO 0x1d00000 (irq = 254, base_baud = 8250000) is a suart_tty
[    2.835050] ti_omapl_pru_suart 1c30000.serial: ti_omapl_pru_suart device registered (pru_clk=228000000, asp_clk=132000000)
[    2.842871] remoteproc remoteproc0: powering up pru1
[    2.842937] remoteproc remoteproc0: Booting fw image ev3-pru1.fw, size 82036
[    2.844681] virtio_rpmsg_bus virtio0: creating channel ev3-tacho-rpmsg addr 0x0
[    2.844863] virtio_rpmsg_bus virtio0: rpmsg host is online
[    2.845119] remoteproc remoteproc0: registered virtio0 (type 7)
[    2.845162] remoteproc remoteproc0: remote processor pru1 is now up
[    2.846356] ev3_pru virtio0.ev3-tacho-rpmsg.-1.0: new channel: 0x400 -> 0x0!
[    2.850552] input: EV3 Brick Buttons as /devices/platform/gpio_keys/input/input1
[    2.857552] ALSA device list:
[    2.857594]   #0: LEGO MINDSTORMS EV3 Speaker connected to 1f00000.pwm
[    2.859114] Freeing unused kernel memory: 184K
[    2.859143] This architecture does not have kernel memory protection.
[    2.859162] Run /init as init process
[    3.115668] usb 1-1: new full-speed USB device number 2 using ohci-da8xx
[    3.393873] usb 1-1: New USB device found, idVendor=050d, idProduct=0121, bcdDevice= 1.01
[    3.393930] usb 1-1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[    3.393964] usb 1-1: Product: USB To  To LAN Conve
[    3.393993] usb 1-1: Manufacturer: ADMtek
[    3.394020] usb 1-1: SerialNumber: 0001
[   11.608024] EXT4-fs (mmcblk0p2): mounted filesystem with ordered data mode. Opts: (null)
[   12.075749] random: crng init done
[   13.351115] systemd[1]: System time before build time, advancing clock.
[   13.653067] systemd[1]: systemd 241 running in system mode. (+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN -PCRE2 default-hierarchy=hybrid)
[   13.657127] systemd[1]: Detected architecture arm.
[   13.671957] systemd[1]: Set hostname to <ev3dev>.
[   18.244821] systemd[1]: File /lib/systemd/system/systemd-journald.service:12 configures an IP firewall (IPAddressDeny=any), but the local system does not support BPF/cgroup based firewalling.
[   18.244959] systemd[1]: Proceeding WITHOUT firewalling in effect! (This warning is only shown for the first loaded unit using IP firewalling.)
[   20.038986] systemd[1]: Listening on Journal Socket.
[   20.068057] systemd[1]: Starting swap with zram...
[   20.133894] systemd[1]: Mounting POSIX Message Queue File System...
[   20.155951] systemd[1]: Listening on initctl Compatibility Named Pipe.
[   20.161753] systemd[1]: Condition check resulted in Huge Pages File System being skipped.
[   20.195169] systemd[1]: Listening on Journal Socket (/dev/log).
[   20.319224] systemd[1]: Starting Create list of required static device nodes for the current kernel...
[   20.330691] systemd[1]: Condition check resulted in Kernel Module supporting RPCSEC_GSS being skipped.
[   20.354949] systemd[1]: Listening on fsck to fsckd communication Socket.
[   20.365204] systemd[1]: Condition check resulted in File System Check on Root Device being skipped.
[   21.247821] zram0: detected capacity change from 0 to 100663296
[   22.217717] RPC: Registered named UNIX socket transport module.
[   22.217750] RPC: Registered udp transport module.
[   22.217764] RPC: Registered tcp transport module.
[   22.217776] RPC: Registered tcp NFSv4.1 backchannel transport module.
[   22.805179] Adding 98300k swap on /dev/zram0.  Priority:16383 extents:1 across:98300k SS
[   23.814975] EXT4-fs (mmcblk0p2): re-mounted. Opts: errors=remount-ro
[   25.309887] systemd[1]: Mounted Kernel Configuration File System.
[   26.663172] systemd[1]: Started Apply Kernel Variables.
[   26.829342] systemd[1]: Started Load/Save Random Seed.
[   27.087121] systemd[1]: Started Create System Users.
[   27.159671] systemd[1]: Starting Create Static Device Nodes in /dev...
[   27.870522] systemd[1]: Started Journal Service.
[   29.903537] systemd-journald[165]: Received request to flush runtime journal from PID 1
[   45.618662] Bluetooth: HCI UART driver ver 2.3
[   45.618710] Bluetooth: HCI UART protocol H4 registered
[   45.618730] Bluetooth: HCI UART protocol BCSP registered
[   45.643133] Bluetooth: HCI UART protocol LL registered
[   45.816291] Bluetooth: hci0: change remote baud rate command in firmware
[   46.079981] pegasus: v0.9.3 (2013/04/25), Pegasus/Pegasus II USB Ethernet driver
[   46.124266] pegasus 1-1:1.0: setup Pegasus II specific registers
[   46.336449] pegasus 1-1:1.0: eth0, Belkin F5D5050 USB Ethernet, 00:05:1b:00:61:c8
[   46.337388] usbcore: registered new interface driver pegasus
[   52.791035] lego-port port2: Added new device 'ev3-ports:in3:ev3-uart-host'
[   52.926183] pegasus 1-1:1.0 enx00051b0061c8: renamed from eth0
[   73.978231] using random self ethernet address
[   73.978279] using random host ethernet address
[   74.559574] using random self ethernet address
[   74.559621] using random host ethernet address
[   74.852846] usb0: HOST MAC 12:16:53:44:cf:21
[   74.857903] usb0: MAC 02:16:53:44:cf:21
[   74.886402] usb1: HOST MAC 32:16:53:44:cf:21
[   74.894605] usb1: MAC 22:16:53:44:cf:21
[   79.754487] IPv6: ADDRCONF(NETDEV_UP): usb0: link is not ready
[   79.874567] IPv6: ADDRCONF(NETDEV_UP): usb1: link is not ready
```

Particularly worrying are these messages:
```
[    0.143711] gpiochip_add_data_with_key: GPIOs 0..143 (davinci_gpio.0) failed to register, -517
[    2.008801] da8xx-usb-phy 1c1417c.chip-controller:usb-phy: Failed to get usb1_clk48
[    2.019386] ehrpwm 1f00000.pwm: Failed to get tbclk
[    2.020123] ehrpwm 1f02000.pwm: Failed to get tbclk
```

**Desktop (please complete the following information):**
 - OS: Xubuntu 18.04.5
 - Connection Type: USB ethernet adapter
 - Other software: SSH

**Robot (please complete the following information):**
 - Device: EV3
 - Versions: [see below - SD card image version, Linux kernel version, etc.]
    - Linux kernel: stock `4.19.85-ev3dev-3-beta2-ev3`, locally-built `4.19.85-ev3dev-3-beta2-ev3-dataage+`
    - SD card image: https://oss.jfrog.org/list/oss-snapshot-local/org/ev3dev/brickstrap/2020-04-10/snapshot-ev3dev-buster-ev3-generic-2020-04-10.img.xz (after unxz-ing, I have md5sum of 045d8184187b1eec1f3a66040dc276bd)

**System info (from `ev3dev-sysinfo`)**
```
Image file:         ev3dev-buster-ev3-generic-2020-04-10
Kernel version:     4.19.85-ev3dev-3-beta2-ev3-dataage+
Brickman:           0.10.2
BogoMIPS:           226.09
Bluetooth:          
Board:              board0
BOARD_INFO_HW_REV=7
BOARD_INFO_MODEL=LEGO MINDSTORMS EV3
BOARD_INFO_ROM_REV=6
BOARD_INFO_SERIAL_NUM=00165344CF21
BOARD_INFO_TYPE=main
```

**Additional context**
I think the SD card may be the problem here, but I cannot prove or disprove that. I think I have used this firmware version successfully previously, but I'm not 100% sure. The brick and the sensors themselves work fine under stock firmware.

**Comment by JakubVanek** [comment]
ev3dev-stretch on the same SD card appears to work - after connecting, the EV3 color sensor turns on its red LED after few seconds.
**System info (from `ev3dev-sysinfo`)**
```
Image file:         ev3dev-stretch-ev3-generic-2020-04-10
Kernel version:     4.14.117-ev3dev-2.3.5-ev3
Brickman:           0.10.3
BogoMIPS:           226.09
Bluetooth:          
Board:              board0
BOARD_INFO_HW_REV=7
BOARD_INFO_MODEL=LEGO MINDSTORMS EV3
BOARD_INFO_ROM_REV=6
BOARD_INFO_SERIAL_NUM=00165344CF21
BOARD_INFO_TYPE=main
```

**Comment by dlech** [comment]
> [    0.143711] gpiochip_add_data_with_key: GPIOs 0..143 (davinci_gpio.0) failed to register, -517

`517` is `EPROBE_DEFER` and is normal. The other errors are probably the same.

**Comment by dlech** [comment]
My first thought is that https://github.com/ev3dev/ev3-systemd/blob/ev3dev-stretch/systemd/ev3-uart%40.service is not working so the line discipline is not being attached.

**Comment by JakubVanek** [comment]
It does look like that might be it, I will try to experiment with it tomorrow.

**Comment by JakubVanek** [comment]
Indeed, `ev3-uart@.service` is the problem: 
```
robot@ev3dev:~$ sudo systemctl list-units --state=failed
  UNIT                             LOAD   ACTIVE SUB    DESCRIPTION                     
● ev3-uart@ev3x2dports:in1.service loaded failed failed LEGO MINDSTORMS EV3 UART Sensors

LOAD   = Reflects whether the unit definition was properly loaded.
ACTIVE = The high-level unit activation state, i.e. generalization of SUB.
SUB    = The low-level unit activation state, values depend on unit type.

1 loaded units listed. Pass --all to see loaded but inactive units, too.
To show all installed unit files use 'systemctl list-unit-files'.
```

```
robot@ev3dev:/lib/systemd/system$ sudo systemctl status ev3-uart@ev3x2dports:in1.service
● ev3-uart@ev3x2dports:in1.service - LEGO MINDSTORMS EV3 UART Sensors
   Loaded: loaded (/lib/systemd/system/ev3-uart@.service; static; vendor preset: enabled)
   Active: failed (Result: exit-code) since Tue 2020-08-11 07:07:38 UTC; 3min 41s ago
  Process: 575 ExecStart=/usr/sbin/ldattach 29 /dev/tty_ev3x2dports:in1 (code=exited, status=1/FAILURE)

Aug 11 07:07:38 ev3dev systemd[1]: Starting LEGO MINDSTORMS EV3 UART Sensors...
Aug 11 07:07:38 ev3dev ldattach[575]: ldattach: cannot open /dev/tty_ev3x2dports:in1: No such file or directory
Aug 11 07:07:38 ev3dev systemd[1]: ev3-uart@ev3x2dports:in1.service: Control process exited, code=exited, status=1/FAILURE
Aug 11 07:07:38 ev3dev systemd[1]: ev3-uart@ev3x2dports:in1.service: Failed with result 'exit-code'.
Aug 11 07:07:38 ev3dev systemd[1]: Failed to start LEGO MINDSTORMS EV3 UART Sensors.
```
The problem seems to be the `-` character encoding. The TTY does exists and running `sudo ldattach 29 /dev/tty_ev3-ports\:in1` successfully initializes the sensor. I tried changing between `%i` and `%I` in the unit file and it had no effect (with `systemctl daemon-reload` after edit).

Contents of `/run/udev/data/+lego\:ev3-ports\:in2\:ev3-uart-host` (I have reattached the sensor) are these:
```
robot@ev3dev:/run/udev/data$ cat +lego\:ev3-ports\:in2\:ev3-uart-host 
I:3253173330
E:SYSTEMD_WANTS=ev3-uart@ev3\x2dports:in2.service
E:SYSTEMD_ALIAS=/sys/subsystem/lego/devices/ev3_uart_host_ev3-ports:in2
G:systemd
```

At this point, I'm starting to feel the pain of people not liking systemd :smile: I tried searching for why the slash might be lost, but I couldn't find anything useful.

**Comment by dlech** [comment]
Maybe caused by this change? https://github.com/systemd/systemd/commit/dcebc9bae4dcc3e844f01558c6127fc0d8745c8e

**Comment by JakubVanek** [comment]
I don't think this change causes it, as this only affects the unit name if it is an empty template. If I use `SYSTEMD_WANT="ev3-uart@.service"`, I get `ev3-uart@sys-devices-platform-ev3\x2dports-ev3\x2dports:in3-lego\x2dport-port2-ev3\x2dports:in3:ev3\x2duart\x2dhost.service` as a full unit name. Unfortunately `SYSTEMD_ALIAS` is not considered for this, only the real sysfs path. At least we get the backslashes back, but the name would require some parsing. The mangling itself at the bottom of the patch doesn't seem to have an effect, as `/bin/systemd-escape --mangle 'ev3-uart@ev3\x2dports:in3.service'` returns the input string with slashes.

**Comment by JakubVanek** [comment]
I think I found the issue:
The `SYSTEMD_WANTS` content is unquoted [here](https://github.com/systemd/systemd/blob/61c8b73a35fd71cb7cbb41b78d93a124bf347e8e/src/core/device.c#L379). The unquoting itself happens [here](https://github.com/systemd/systemd/blob/f1cc283a5ab47aae64a2bdb4cd7f1916309906ff/src/basic/extract-word.c#L19). As only `EXTRACT_UNQUOTE` is specified, the backslash is discarded as no special unescaping mode is specified. Duplicating the backslash should work the problem around.

**Comment by JakubVanek** [comment]
Workaround confirmed. By hardcoding this to `60-ev3.rules`, the port 4 got successfully activated:
```udev
SUBSYSTEMS=="lego", ENV{DEVTYPE}=="ev3-uart-host", ACTION=="add", TAG+="systemd", \
    ENV{SYSTEMD_WANTS}="ev3-uart@ev3\\x2dports:in4.service", \
    ENV{SYSTEMD_ALIAS}="/sys/subsystem/lego/devices/ev3_uart_host_$env{LEGO_ADDRESS}"
```

**Comment by JakubVanek** [comment]
I have created https://github.com/systemd/systemd/issues/16735 to see if this is really the issue.

**Comment by JakubVanek** [comment]
I have found a workaround that wouldn't require backporting the systemd fix:
`/lib/systemd/system/ev3-uart@.service`:
```
# This file is part of the ev3-systemd package

[Unit]
Description=LEGO MINDSTORMS EV3 UART Sensors
# This is the alias created by the ev3dev-uart udev rule.
BindsTo=%i.device
Wants=systemd-udev-settle.service
After=systemd-udev-settle.service

[Service]
Type=forking
ExecStart=/usr/share/ev3-systemd/ev3-start-uart.sh %I
```
part of `/lib/udev/rules.d/60-ev3.rules`:
```
# Start EV3 UART sensor line discipline when UART sensor is connected
SUBSYSTEMS=="lego", ENV{DEVTYPE}=="ev3-uart-host", ACTION=="add", TAG+="systemd", \
    ENV{SYSTEMD_WANTS}="ev3-uart@.service"
```
`/usr/share/ev3-systemd/ev3-start-uart.sh`
```
#!/bin/dash

# Line discipline helper for systemd
#
# Copyright (C) 2020 Jakub Vanek <linuxtardis@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, see <http://www.gnu.org/licenses/>.

set -e

# command line parameters
sysfs_device="$1" # sysfs device path like sys/devices/platform/ev3-ports/ev3-ports:in3/lego-port/port2/ev3-ports:in3:ev3-uart-host
device_name="$(echo "$sysfs_device" | sed 's|:ev3-uart-host||g; s|sys/devices/platform/ev3-ports/ev3-ports:in./lego-port/port./||g')"
exec /usr/sbin/ldattach 29 "/dev/tty_$device_name"
```

I tried to add the second backslash in the udev rule, but the preceding backslashes would be always replaced by `_`.

**Comment by JakubVanek** [comment]
The behaviour of systemd turned out to be correct. The `SYSTEMD_WANTS` value in udev can be enclosed in quotes and for this reason systemd eats the backslashes, marking the "escaped" characters as non-quotes.

On the other hand, there might another feature missing in udev and that is the ability to suspend invalid escape sequence replacement for `PROGRAM` output. This would allow us to pass `\\` using a sed script, but this seems not to be possible. `OPTIONS+="string_escape=none"` ought to enable this in v241 udev, but it doesn't work and v243+ removes the code that would have allowed this.

The proposed workaround may be the only viable solution then, as I don't think I want to backport new udev features to the stable debian version. It will require splitting buster and stretch versions of the `ev3-systemd` package, as systemd in stretch (v232) does not contain https://github.com/systemd/systemd/commit/dcebc9bae4dcc3e844f01558c6127fc0d8745c8e.

**Comment by JakubVanek** [comment]
I have prepared a fixed package at https://github.com/JakubVanek/ev3-systemd-private-dev/commits/ev3dev-buster . It fixes this bug by introducing the workaround and also it updates the packaging. Regarding copyright: I don't care about where it goes, I just felt that writing other's name as being responsible for the file and consequently all its errors is not the best thing. Feel free to unify it so that the package has only one copyright holder.

**Comment by dlech** [comment]
I've actually been thinking about skipping ev3dev-buster and going straight to ev3dev-bullseye since bullseye will probably become stable in the next year or so. Would this help since bullseye has a newer version of systemd.?

**Comment by JakubVanek** [comment]
I don't think so, current behaviour on buster corresponds to how systemd `master` would behave from looking into the source code. However, bullseye could still receive some new code to pass the double backslash from udev to systemd.

**Comment by dlech** [comment]
I have published your changes to ev3dev-buster.

**Comment by dlech** [comment]
doh, and I just accidentally deleted it

**Comment by dlech** [comment]
and it is back again now

**Comment by JakubVanek** [comment]
Thank you, I can confirm that running `sudo apt-get update && sudo apt-get dist-upgrade` on clean 2020-04-10 ev3dev-buster makes UART sensors work again.

**Comment by CarloWood** [comment]
Why was this closed? There is no way I can pass an escaped character with `SYSTEMD_WANTS`, and I tried *everything* in the past three hours.

My last attempt looks like this:
```
SUBSYSTEMS=="block", ENV{DEVTYPE}=="partition", ENV{ID_SERIAL_SHORT}=="408D5CBECBBDE7C0E9160666|408D5CBF5F0AE7C0E9150B70|408D5CBECAC3E7C0E9170AEC", PROGRAM="/usr/bin/systemd-escape 'foo-bar'", RUN+="/usr/local/sbin/usbtest '%c'", ENV{SYSTEMD_WANTS}="B@$result.service"
```
This results in `/usr/local/sbin/usbtest` being passed the correct value (`foo\x2dbar`) while the service `B@foox2dbar.service` is being started (removed backslash).

I tried adding backslashes, different quotes, using $result etc etc. Absolutely no way to get the right service started.

Doing `sudo systemctl start 'B@foo\x2dbar.service'` from the command line works as expected, so the bug is clearly with udev and SYSTEMD_WANTS.

Is there a solution for this?!

Using version 257
